librosa/core/constantq.py

```diff
@@ -612,115 +604,80 @@ def icqt(C, sr=22050, hop_length=512, fmin=None,
-                                        pad_fft=True)
-    n_fft = basis.shape[1]
+    fmin_t = np.min(freqs[-bins_per_octave:])
 
-    # The extra factor of lengths**0.5 corrects for within-octave tapering
-    basis = basis * np.sqrt(lengths[:, np.newaxis])
+    n_filters = min(n_bins, bins_per_octave)
 
-    # Estimate the gain per filter
-    bdot = basis.conj().dot(basis.T)
-    bscale = np.sum(np.abs(bdot), axis=1)
+    fft_basis, n_fft, lengths = __cqt_filter_fft(sr, fmin_t,
+                                                 n_filters,
+                                                 bins_per_octave,
+                                                 tuning,
+                                                 filter_scale,
+                                                 norm,
+                                                 sparsity=sparsity,
+                                                 window=window)
 
-    n_trim = basis.shape[1] // 2
+    if min(lengths) < hop_length:
+        warnings.warn('Minimum CQT filter length={:.3f} is less than hop_length={}.\n'
+                      'This will probably cause unpleasant acoustic artifacts. '
+                      'Consider decreasing your hop length or increasing the frequency resolution of your CQT.'.format(min(lengths), hop_length))
 
-    if scale:
-        Cnorm = np.ones(n_bins)[:, np.newaxis]
-    else:
-        Cnorm = filters.constant_q_lengths(sr=sr,
-                                           fmin=fmin,
-                                           n_bins=n_bins,
-                                           bins_per_octave=bins_per_octave,
-                                           filter_scale=filter_scale,
-                                           tuning=tuning,
-                                           window=window)[:, np.newaxis]**0.5
+    # The basis gets renormalized by the effective window length above;
+    # This step undoes that
+    fft_basis = fft_basis.todense() * n_fft / lengths[:, np.newaxis]
 
-    y = None
+    # This step conjugate-transposes the filter
+    inv_basis = np.asanyarray(fft_basis).conj().T
 
-    # Revised algorithm:
-    #   for each octave
-    #      upsample old octave
-    #      @--numba accelerate this loop?
-    #      for each basis
-    #         convolve with activation (valid-mode)
-    #         divide by window sumsquare
-    #         trim and add to total
+    # How many octaves do we have?
+    n_octaves = int(np.ceil(float(n_bins) / bins_per_octave))
 
+    y = None
     for octave in range(n_octaves - 1, -1, -1):
-        # Compute the slice index for the current octave
         slice_ = slice(-(octave+1) * bins_per_octave - 1,
                        -(octave) * bins_per_octave - 1)
 
-        # Project onto the basis
-        C_oct = C[slice_] / Cnorm[slice_]
-        basis_oct = basis[-C_oct.shape[0]:]
-
-        y_oct = None
+        # Slice this octave
+        C_oct = C[slice_]
```

Comment:
not sure what's going on here. why is there only a single subscript if `C` is 2-D? is it something that is taken into account by the `slice` Python function? also, isn't `-C_oct.shape[0]` a constant (independent from `octave`)? why does it need to be estimated at runtime? in any case, this part deserves more comment than the laconic `Slice this octave` :)